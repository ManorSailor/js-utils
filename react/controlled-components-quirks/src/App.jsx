import InputHandlerOnly from "./components/Input/WithHandlerOnly";
import InputWithNullValue from "./components/Input/WithNullValue";
import InputValueOnly from "./components/Input/WithValueOnly";
import Label from "./components/Label";

function App() {
  return (
    <>
      <Label labelName={InputHandlerOnly.name}>
        <InputHandlerOnly />
      </Label>

      <Label labelName={InputValueOnly.name}>
        <InputValueOnly />
      </Label>

      <Label labelName={InputWithNullValue.name}>
        <InputWithNullValue />
      </Label>
    </>
  );
}

export default App;
