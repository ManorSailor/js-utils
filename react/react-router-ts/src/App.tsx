import { ReactElement, useEffect, useState } from "react";
import { Route, Routes } from "react-router-dom";

import Navbar from "@components/Navbar";
import { Home, Product, Products, NotFound } from "./pages";

function App(): ReactElement {
  const [products, setProducts] = useState<Product[]>([]);

  useEffect(() => {
    fetch("https://dummyjson.com/products?limit=3")
      .then((res) => res.json())
      .then(({ products }: DummyJSONResponse) => setProducts(products))
      .catch((err: Error) => err);
  }, []);

  return (
    <>
      <Navbar />
      <Routes>
        <Route path="/" element={(<Home />) as ReactElement} />
        <Route path="/products">
          <Route
            index
            element={(<Products products={products} />) as ReactElement}
          />
          <Route
            path=":id"
            element={(<Product products={products} />) as ReactElement}
          />
        </Route>
        <Route path="*" element={<NotFound />} />
      </Routes>
    </>
  );
}

export default App;
