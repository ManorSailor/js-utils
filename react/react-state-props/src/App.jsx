import React, { Component } from 'react';

import Counter from './components/Counter';
import MultiplyClass from './components/MultiplyClass/Multiply';
import MultiplyHooks from './components/MultiplyHooks/Multiply';
import Lifecycle from './components/LifecycleCounter/Lifecycle';
import ColorHooks from './components/ColorHooks/ColorHooks';

class App extends Component {
  state = {
    count: 1,
  };

  countUp = () => this.setState({ count: this.state.count + 1 });
  countDown = () => this.setState({ count: this.state.count - 1 });

  render() {
    const { count } = this.state;

    return (
      <div className="App">
        <Counter
          count={count}
          countUp={this.countUp}
          countDown={this.countDown}
        />

        <MultiplyClass multiplicand={count} />
        <MultiplyHooks multiplicand={count} />
        <ColorHooks />
        <Lifecycle />
      </div>
    );
  }
}

export default App;
