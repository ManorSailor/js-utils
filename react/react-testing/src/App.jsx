import { useState } from 'react';

function App() {
  const [header, setHeader] = useState('Our First Test');

  return <h1 onClick={() => setHeader('Our Second Test')}>{header}</h1>;
}

export default App;
