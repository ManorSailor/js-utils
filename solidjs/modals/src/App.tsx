import { type Component } from "solid-js";
import { ConfirmationModal } from "./components/ConfirmationModal";
import { useConfirmationModal } from "./hooks/useConfirmationModal";

const App: Component = () => {
  const { showModal, ...props } = useConfirmationModal();

  return (
    <>
      <button onclick={showModal}>Show Modal</button>

      <ConfirmationModal {...props} />
    </>
  );
};

export default App;
