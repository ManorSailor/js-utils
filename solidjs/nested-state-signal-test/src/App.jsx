import { createEffect, createSignal, onCleanup } from "solid-js";

// State is not tied to components unlike React. By tied I meant, it cannot exist outside of components
const [state, setState] = createSignal({
  compA: 0,
  compB: 0,
});

// Similar pattern we all have seen before a thousand times
const changeState = (key, val) =>
  setState((oldState) => ({ ...oldState, [key]: val }));

// Change only one of the states. value of the state doesn't matter
// Whether a state consumer NOT using `compA` will be notified or not
setInterval(() => changeState("compA", "ha"), 1000);

// Finally, track states for changes. This consumer is consuming a piece of state which does not changes. Will it run?
createEffect(() => console.log("B renders:", state().compB));

// Conclusion: Yes. When `compA` changes, `compB` dependent is also notified even though it doesn't rely on `compA` at all.
// Why? Because I believe signals perform a Strict Equality check on new & old state, & objects are compared by references.
// And if you look at changeState, we give it a new object each time copying the old values.
//
// Fix? Well, it's not really an issus to be fixed. The real issue lies that you are using wrong tool for the job.
// Use `Stores` if you really want that fine-grained control over your state. They (lazily & auto-magically) create a signal for each property of the object

export default function App() {}
