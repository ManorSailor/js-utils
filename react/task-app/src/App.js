import './App.css';
import React from 'react';
import TaskList from './components/TaskList';
import uniqid from 'uniqid';

class App extends React.Component {
  state = {
    tasks: [],
    task: { title: '', id: uniqid(), order: 1 },
  };

  handleChange(e) {
    const { task } = this.state;

    this.setState({
      task: {
        ...task,
        title: e.target.value,
      },
    });
  }

  onSubmitTask(e) {
    const { task, tasks } = this.state;

    this.setState({
      tasks: [...tasks, task],
      task: {
        title: '',
        id: uniqid(),
        order: task.order + 1,
      },
    });

    e.target.reset();
    e.preventDefault();
  }

  deleteTask(removedTask) {
    const { tasks } = this.state;

    this.setState({
      tasks: tasks.filter((task) => task !== removedTask),
    });
  }

  updateTask(updatedTask) {
    const { tasks } = this.state;

    this.setState({
      tasks: tasks.map((task) =>
        task.id === updatedTask.id ? updatedTask : task
      ),
    });

    // Is the approach below okay? It works & updates the task.
    // Pros: Doesn't cause a re-render
    // Cons: Too specific, works only if the title was updated; Mutating state?
    // const task = this.state.tasks.find((task) => task.id === updatedTask.id);
    // task.title = updatedTask.title;
  }

  render() {
    const { task, tasks } = this.state;

    return (
      <>
        <form onSubmit={this.onSubmitTask.bind(this)}>
          <label htmlFor="task-field">
            Enter Task
            <input
              type="text"
              id="task-field"
              value={task.title}
              onChange={this.handleChange.bind(this)}
            />
          </label>
          <button>Add Task</button>
        </form>

        <TaskList
          tasks={tasks}
          deleteTask={this.deleteTask.bind(this)}
          updateTask={this.updateTask.bind(this)}
        />
      </>
    );
  }
}

export default App;
