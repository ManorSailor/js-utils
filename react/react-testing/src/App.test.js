import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import App from './App';

describe('App Component', () => {
  it('renders initial heading', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('changes heading on click', async () => {
    const user = userEvent.setup();

    render(<App />);
    const header = screen.getByRole('heading', { name: /our first test/i });

    await user.click(header);

    expect(header).toHaveTextContent(/our second test/i);
  });
});
