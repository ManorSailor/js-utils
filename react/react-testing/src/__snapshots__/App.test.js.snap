// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component renders initial heading 1`] = `
<div>
  <h1>
    Our First Test
  </h1>
</div>
`;
